test/metabase/data_apps/api/app_test.clj

```diff
@@ -0,0 +1,99 @@
+(ns metabase.data-apps.api.app-test
+  "Tests for /api/app endpoints for data app consumers"
+  (:require
+   [clojure.test :refer :all]
+   [metabase.data-apps.test-util :as data-apps.tu]
+   [metabase.test :as mt]
+   [ring.util.codec :as codec]
+   [toucan2.core :as t2]))
+
+(defn- get-app-slug
+  "Helper function to construct a properly slug-encoded app endpoint."
+  [app-slug]
+  (str "/app/" (codec/url-encode app-slug)))
+
+(defn- get-app-list
+  "Helper function to fetch the list of apps."
+  ([user status query-params]
+   (let [endpoint (if query-params
+                    (str "/app/?" query-params)
+                    "/app/")]
+     (mt/user-http-request user :get status endpoint)))
+  ([]
+   (get-app-list :crowberto 200 nil)))
+
+(defn- filter-apps-by-ids
+  [response app-ids]
+  (assert (set? app-ids))
+  (assoc response :data (filter #(app-ids (:id %)) (:data response))))
+
+(deftest get-published-apps-test
+  (testing "GET /api/app/ - fetch published data apps"
+    (data-apps.tu/with-released-app!
+      [app {}]
+      (is (=? {:data [{:id (:id app)}]
+               :limit nil
+               :offset nil
+               :total (mt/malli=? pos-int?)}
+              (filter-apps-by-ids (get-app-list)
+                                  #{(:id app)})))))
+
+  (testing "GET /api/app/ - pagination works"
+    (data-apps.tu/with-released-app!
+      [app1 {:name "App 1" :slug "/app1"}]
+      (data-apps.tu/with-released-app!
+        [app2 {:name "App 2" :slug "/app2"}]
+        (let [response (filter-apps-by-ids (get-app-list :crowberto 200 "limit=1&offset=0")
+                                           #{(:id app1) (:id app2)})]
+          (is (=? {:data [{:id (:id app2)}]
+                   :limit 1
+                   :offset 0
+                   :total (mt/malli=? pos-int?)}
+                  response)))))))
+
+(deftest get-published-app-by-slug-test
```

Comment:
What about a test for the case where the status is `published` but there are no versions?

I guess this should also be a 404, although we might want to prevent getting into this state.

Reply:
Test added